constantine/math/arithmetic/assembly/limbs_asm_bigint_arm64.nim

```diff
@@ -0,0 +1,212 @@
+    r = asmArray(r_PIR, N, PointerInReg, asmInput, memIndirect = memWrite)
+    a = asmArray(a_PIR, N, PointerInReg, asmInput, memIndirect = memRead)
+    b = asmArray(b_PIR, N, PointerInReg, asmInput, memIndirect = memRead)
+
+    u0Sym = ident"u0"
+    u1Sym = ident"u1"
+    v0Sym = ident"v0"
+    v1Sym = ident"v1"
+
+  var # Swappable registers to break dependency chains
+    u0 = asmValue(u0Sym, Reg, asmOutputEarlyClobber)
+    u1 = asmValue(u1Sym, Reg, asmOutputEarlyClobber)
+    v0 = asmValue(v0Sym, Reg, asmOutputEarlyClobber)
+    v1 = asmValue(v1Sym, Reg, asmOutputEarlyClobber)
+
+  # Prologue
+  result.add quote do:
+    var `u0sym`{.noinit.}, `u1sym`{.noinit.}: BaseType
+    var `v0sym`{.noinit.}, `v1sym`{.noinit.}: BaseType
+
+  # Algorithm
+  if N >= 2:
+    ctx.ldp u0, u1, a[0]
+    ctx.ldp v0, v1, b[0]
+  else:
+    ctx.ldr u0, a[0]
+    ctx.ldr v0, b[0]
+
+  for i in 0 ..< N:
+    if i == 0:
+      ctx.adds u0, u0, v0
+    else:
+      ctx.adcs u0, u0, v0
+    ctx.str u0, r[i]
+
+    # Next iteration
+    if i != N-1:
+      swap(u0, u1)
+      swap(v0, v1)
+      if i+2 < N:
+        ctx.ldr u1, a[i+2]
+        ctx.ldr v1, b[i+2]
+
+  ctx.setOutputToFlag(carry, CarryFlag)
+
+  # Codegen
+  result.add ctx.generate()
+
+func add_asm*(r: var Limbs, a, b: Limbs): Carry =
+  ## Constant-time addition
+  add_gen(result, r, a, b)
+
+# Subtraction
+# ------------------------------------------------------------
+
+macro sub_gen[N: static int](borrow: var Borrow, r_PIR: var Limbs[N], a_PIR, b_PIR: Limbs[N]): untyped =
+  ## Generate an optimized out-of-place subtraction kernel
+  result = newStmtList()
+
+  var ctx = init(Assembler_arm64, BaseType)
+  let
+    # MemOffsettable is the better constraint but
+    # with ARM64 we cannot generate array offsets from it due to inline ASM auto-bracketings
+    r = asmArray(r_PIR, N, PointerInReg, asmInput, memIndirect = memWrite)
+    a = asmArray(a_PIR, N, PointerInReg, asmInput, memIndirect = memRead)
+    b = asmArray(b_PIR, N, PointerInReg, asmInput, memIndirect = memRead)
+
+    u0Sym = ident"u0"
+    u1Sym = ident"u1"
+    v0Sym = ident"v0"
+    v1Sym = ident"v1"
+
+  var # Swappable registers to break dependency chains
+    u0 = asmValue(u0Sym, Reg, asmOutputEarlyClobber)
+    u1 = asmValue(u1Sym, Reg, asmOutputEarlyClobber)
+    v0 = asmValue(v0Sym, Reg, asmOutputEarlyClobber)
+    v1 = asmValue(v1Sym, Reg, asmOutputEarlyClobber)
+
+  # Prologue
+  result.add quote do:
+    var `u0sym`{.noinit.}, `u1sym`{.noinit.}: BaseType
+    var `v0sym`{.noinit.}, `v1sym`{.noinit.}: BaseType
+
+  # Algorithm
+  if N >= 2:
+    ctx.ldp u0, u1, a[0]
+    ctx.ldp v0, v1, b[0]
+  else:
+    ctx.ldr u0, a[0]
+    ctx.ldr v0, b[0]
+
+  for i in 0 ..< N:
+    if i == 0:
+      ctx.subs u0, u0, v0
+    else:
+      ctx.sbcs u0, u0, v0
+    ctx.str u0, r[i]
+
+    # Next iteration
+    if i != N-1:
+      swap(u0, u1)
+      swap(v0, v1)
+      if i+2 < N:
+        ctx.ldr u1, a[i+2]
+        ctx.ldr v1, b[i+2]
```

Comment:
For now this uses fancy prefetching but unsure if beneficial on Apple Silicon (can fetch/decode up to 8 instructions per cycle) or Raspberry Pi 5